packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/PropertiesGrouping.ts

```diff
@@ -0,0 +1,332 @@
+    let groupingNodeKey: PropertyGroupingNodeKey = {
+      type: "other-property-grouping",
+      groupingInfo: {
+        fullClassName: entry.property.fullClassName,
+        propertyName: entry.property.propertyName,
+      },
+    };
+    const extraValues = entry.property.extraValues;
+    if (extraValues) {
+      if ("formattedValue" in extraValues) {
+        groupingNodeKey = {
+          type: "formatted-property-grouping",
+          groupingInfo: {
+            fullClassName: entry.property.fullClassName,
+            propertyName: entry.property.propertyName,
+            formattedPropertyValue: extraValues.formattedValue,
+          },
+        };
+      } else {
+        groupingNodeKey = {
+          type: "ranged-property-grouping",
+          groupingInfo: {
+            fullClassName: entry.property.fullClassName,
+            propertyName: entry.property.propertyName,
+            fromValue: extraValues.fromValue,
+            toValue: extraValues.toValue,
+          },
+        };
+      }
+    }
+    const groupedNodeParentKeys = entry.groupedNodes[0].parentKeys;
+    outNodes.grouped.push({
+      label: entry.property.label,
+      key: groupingNodeKey,
+      parentKeys: groupedNodeParentKeys,
+      children: entry.groupedNodes.map((gn) => ({ ...gn, parentKeys: [...groupedNodeParentKeys, groupingNodeKey] })),
+    });
+  });
+  outNodes.ungrouped.push(...groupings.ungrouped);
+  return outNodes;
+}
+
+/** @internal */
+export async function getUniquePropertiesGroupInfo(metadata: IMetadataProvider, nodes: ProcessedInstanceHierarchyNode[]): Promise<Array<PropertyGroupInfo>> {
+  const uniqueProperties = new Map<string, PropertyGroupInfo>();
+  for (const node of nodes) {
+    const byProperties = node.processingParams?.grouping?.byProperties;
+    if (!byProperties) {
+      continue;
+    }
+
+    const previousPropertiesInfo = new Array<{ propertyGroup: PropertyGroup; propertyGroupKey: string }>();
+    for (const propertyGroup of byProperties.propertyGroups) {
+      const lastKey = previousPropertiesInfo.length > 0 ? `[${previousPropertiesInfo[previousPropertiesInfo.length - 1].propertyGroupKey}]` : "[]";
+
+      const mapKeyRanges = getRangesAsString(propertyGroup.ranges);
+      const mapKey = `${byProperties.fullClassName}:${lastKey}:${propertyGroup.propertyName}:${mapKeyRanges}`;
+      if (uniqueProperties.get(mapKey)) {
+        addToPreviousPropertiesInfo(previousPropertiesInfo, propertyGroup);
+        continue;
+      }
+      uniqueProperties.set(mapKey, {
+        ecClass: await getClass(metadata, byProperties.fullClassName),
+        propertyGroup: {
+          propertyName: propertyGroup.propertyName,
+          ranges: propertyGroup.ranges,
+        },
+        previousPropertiesGroupingInfo: [
+          ...previousPropertiesInfo.map((groupingInfo) => {
+            return { fullClassName: byProperties.fullClassName, propertyGroup: groupingInfo.propertyGroup };
+          }),
+        ],
```

Comment:
As I understand, you are suggesting to remove `const previousPropertiesInfo = new Array<{ propertyGroup: PropertyGroup; propertyGroupKey: string }>();` and instead save `propertyGroupKey` in the `previousPropertiesGroupingInfo`. We can do that, but I am just wondering if it is correct to save the keys there, since this  `previousPropertiesGroupingInfo` is also used `createPropertyGroups()` function.

Reply:
This is 'previousPropertiesGroupingInfo' is also used in the 'createPropertyGroups' so I don't think that we should store map keys there.